assets/model_monitoring/components/src/token_statistics_compute_metrics/compute_token_statistics_metrics.py

```diff
@@ -0,0 +1,351 @@
+    if null_id_count != 0:
+        print(f"Error: token_df contains {null_id_count} null values in id column")
+        print(f"filtering out the rows with null id")
+        # filter out the rows with null id
+        token_df = token_df.filter(token_df["id"].isNotNull())
+    
+    null_node_id_count =  token_df.filter(token_df["node_id"].isNull()).count()
+    if null_node_id_count != 0:
+        print(f"Error: token_df contains {null_node_id_count} null values in node_id column")
+        print(f"filtering out the rows with null node_id")
+        # filter out the rows with null node_id
+        token_df = token_df.filter(token_df["node_id"].isNotNull())
+
+    # check that for rows where status_code is 200, no expected column has a null value. If so, filter out those rows
+    for column in expected_columns:
+        if column == "status_code":
+            continue
+        null_count = token_df.filter(token_df["status_code"] == 200).filter(
+            token_df[column].isNull()
+        ).count()
+        if null_count != 0:
+            print(f"Error: token_df contains {null_count} null values in {column} column for rows where status_code is 200")
+            print(f"filtering out the rows with null {column}")
+            # filter out the rows with null column
+            token_df = token_df.filter(token_df[column].isNotNull())
+    return token_df
+
+def compute_conditional_counts_df(token_df, dimensions, condition_column, condition_value, metric_name):
+    '''
+    Needs the token_df with columns dimensions, condition_column
+    Returns a metric df with columns group, group_pivot, metric_name, metric_value
+    '''
+    # check if the token_df has the expected columns
+    if set(dimensions + [condition_column]).issubset(token_df.columns) == False:
+        return
+    
+    return token_df.filter(token_df[condition_column] == condition_value).groupBy(dimensions).agg(
+    lit(metric_name).alias('metric_name'),
+    count('*').cast("float").alias("metric_value"),
+    )
+
+def compute_avg_df(token_df, columns, dimensions, metric_prefix=""):
+    '''
+    Needs the token_df with columns dimensions, columns
+    Returns a metric df with columns group, group_pivot, metric_name, metric_value
+    where metric_name is the column name and metric_value is the average of the column
+    '''
+    # check if the token_df has the expected columns
+    if set(dimensions + columns + ['status_code']).issubset(token_df.columns) == False:
+        return
+    # restrict the token_df to only expected columns
+    token_df = token_df.select(dimensions + columns + ['status_code'])
+    
+    avg_columns = [avg(col).cast("float").alias(f"{metric_prefix}_avg_{col}") for col in columns]
+    stack_columns = ["'"+ metric_prefix +"_avg_" + col + "', " + metric_prefix + "_avg_" + col for col in columns]
+    stack_expr = "stack(" + str(len(columns)) + ", " + ", ".join(stack_columns) + ") as (metric_name, metric_value)"
```

Comment:
suggestion:  A one-liner explanation with a sample for this expression will be tremendously helpful to those are new to data science like me to digest.